src/mono/mono/mini/method-to-ir.c

```diff
@@ -9844,14 +9844,22 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 
 			context_used = mini_class_check_context_used (cfg, klass);
 
-			if (sp [0]->type == STACK_I8 || (TARGET_SIZEOF_VOID_P == 8 && sp [0]->type == STACK_PTR)) {
+			if (sp [0]->type == STACK_I8 && TARGET_SIZEOF_VOID_P == 4) {
 				MONO_INST_NEW (cfg, ins, OP_LCONV_TO_OVF_U4);
```

Comment:
Why is this needed ? It works on every other platform.

Reply:
Shouldn't the OP_LCONV_TO_OVF_U4 implementation on s390x be changed to clear the upper bits like the other platforms do ?

Reply:
OP_LCONV_TO_OVF_U4 *does* clear upper bits on s390x (although I think this should not be necessary since normal convention on s390x is to store 32-bit values in 64-bit GPRs with the upper bits undefined).

However, in the case where I ran into this bug originally, the OP_LCONV_TO_OVF_U4 was not even emitted, because the incoming value was neither a STACK_I8 *nor* a STACK_PTR, but rather a STACK_I4.

This meant this condition was skipped completely and the STACK_I4 directly passed into ALLOC_VECTOR, which expects a STACK_I8 (or STACK_PTR I guess).